### PR TITLE
K8SPSMDB-333 Test migration from RS to shards

### PR DESCRIPTION
**Pull request by cap1984**
[K8SPSMDB-333](https://jira.percona.com/browse/K8SPSMDB-333)

Basic test case for a common user

### CODE REVIEW

File: e2e-tests/rs-shard-migration/run [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    sleep 10
+    wait_for_running "${cluster}-rs0" "${CLUSTER_SIZE}" "false"
+    wait_for_running "${cluster}-cfg" "${CLUSTER_SIZE}" "false"
+    wait_cluster_consistency "${cluster}" "${CLUSTER_SIZE}"
+    # Migration to shards wipes out rs users. Let's recreate them
+    run_mongos 'db.createUser({user: "myApp", pwd: "myPass", roles: [{ db: "myApp", role: "readWrite" }]})' \
+                "userAdmin:userAdmin123456@${cluster}-mongos.${namespace}"
+
+    simple_data_check "${cluster}" "${CLUSTER_SIZE}" 1 "-mongos"
+
```
Comment (delgod) [inline e2e-tests/rs-shard-migration/run:33]: I believe we need to add the backward conversion to replicaset

**Review by delgod — CHANGES_REQUESTED** [review]

File: e2e-tests/rs-shard-migration/run (hunk 1 above)
Comment (cap1984) [reply e2e-tests/rs-shard-migration/run:33]: ок

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-497-a99c54c0

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| rs-shard-migration | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| users | passed |
| upgrade-sharded | passed |

File: e2e-tests/rs-shard-migration/run (hunk 1 above)
Comment (cap1984) [reply e2e-tests/rs-shard-migration/run:33]: done

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| scheduled-backup | failed |
| operator-self-healing | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| demand-backup-sharded | failed |
| users | passed |
| upgrade | passed |
| upgrade-sharded | passed |

File: e2e-tests/rs-shard-migration/run [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    kubectl_bin patch psmdb/${cluster} --type json -p='[{"op":"add","path":"/spec/sharding","value":{"configsvrReplSet":{"size":'${CLUSTER_SIZE}',"volumeSpec":{"persistentVolumeClaim":{"resources":{"requests":{"storage":"3Gi"}}}}},"enabled":true,"mongos":{"size":1,"expose":{"exposeType":"ClusterIP"}}}}]'
+    sleep 10
+    wait_for_running "${cluster}-rs0" "${CLUSTER_SIZE}" "false"
+    wait_for_running "${cluster}-cfg" "${CLUSTER_SIZE}" "false"
+    wait_cluster_consistency "${cluster}" "${CLUSTER_SIZE}"
+    # Migration to shards wipes out rs users. Let's recreate them
+    run_mongos 'db.createUser({user: "myApp", pwd: "myPass", roles: [{ db: "myApp", role: "readWrite" }]})' \
+                "userAdmin:userAdmin123456@${cluster}-mongos.${namespace}"
+
+    simple_data_check "${cluster}" "${CLUSTER_SIZE}" 1 "-mongos"
```
Comment (delgod) [inline e2e-tests/rs-shard-migration/run:32]: can we connect to `-rs0` and run command `db.isMaster().lastCommittedOpTime` (or `db.isMaster()` and check `$configServerState` field
1) if sharding enabled we should check if lastCommittedOpTime is non empty
2) if sharding disabled we should check if lastCommittedOpTime is empty

**Review by delgod — CHANGES_REQUESTED** [review]

File: e2e-tests/rs-shard-migration/run (hunk 2 above)
Comment (cap1984) [reply e2e-tests/rs-shard-migration/run:46]: done

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-497-4455db58

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| arbiter | passed |
| rs-shard-migration | passed |
| demand-backup-sharded | passed |
| service-per-pod | passed |
| liveness | passed |
| upgrade | passed |
| smart-update | passed |
| upgrade-sharded | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-497-6465d55e

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| upgrade | passed |
| version-service | passed |
| upgrade-sharded | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| arbiter | passed |
| scaling | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| upgrade | passed |
| version-service | failed |
| upgrade-sharded | passed |
| users | passed |

File: e2e-tests/rs-shard-migration/run [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    local isSharded=$3
+
+    if [[ ${isSharded} == true ]]; then
+        echo $(run_mongos 'db.isMaster().lastCommittedOpTime' "clusterAdmin:clusterAdmin123456@${cluster_name}-rs0-0.${cluster_name}-rs0.${namespace}" "mongodb" "" "--quiet" | egrep -v 'I NETWORK|W NETWORK|Error saving history file|Percona Server for MongoDB|connecting to:|Unable to reach primary for set|Implicit session:|versions do not match')
+        return
+    else
+        echo $(run_mongo 'db.isMaster().lastCommittedOpTime' "clusterAdmin:clusterAdmin123456@${cluster_name}-rs0-0.${cluster_name}-rs0.${namespace}" "mongodb" "" "--quiet" | egrep -v 'I NETWORK|W NETWORK|Error saving history file|Percona Server for MongoDB|connecting to:|Unable to reach primary for set|Implicit session:|versions do not match')
+        return
+    fi
+}
```
Comment (delgod) [inline e2e-tests/rs-shard-migration/run:21]: keep it simple
```
function get_lastCommitedOpTime() {
    local cluster_name=$1
    local namespace=$2
    run_mongo 'db.isMaster().lastCommittedOpTime' "clusterAdmin:clusterAdmin123456@${cluster_name}-rs0-0.${cluster_name}-rs0.${namespace}" "mongodb" "" "--quiet" \
        | egrep -v 'I NETWORK|W NETWORK|Error saving history file|Percona Server for MongoDB|connecting to:|Unable to reach primary for set|Implicit session:|versions do not match'
}
```

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-497-dc8b1a02

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| upgrade | passed |
| version-service | passed |
| upgrade-sharded | passed |
| users | passed |

File: e2e-tests/rs-shard-migration/run (hunk 3 above)
Comment (cap1984) [reply e2e-tests/rs-shard-migration/run]: done

**Review by delgod — DISMISSED** [review]

**Review by tplavcic — CHANGES_REQUESTED** [review]
Please check mongos deployment status after we migrate to sharding and also when we migrate back to replica that mongos and cfg services are removed.

**Comment by cap1984** [comment]
> Please check mongos deployment status after we migrate to sharding and also when we migrate back to replica that mongos and cfg services are removed.

done

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | failed |
| demand-backup | passed |
| self-healing | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| operator-self-healing | passed |
| arbiter | passed |
| rs-shard-migration | passed |
| service-per-pod | failed |
| scheduled-backup | failed |
| demand-backup-sharded | failed |
| upgrade | failed |
| liveness | failed |
| upgrade-sharded | failed |

File: e2e-tests/rs-shard-migration/run [hunk 4]
```diff
@@ -50,7 +55,8 @@ function main() {
 
     simple_data_check "${cluster}-rs0" "${CLUSTER_SIZE}"
 
-    if [[ ! -z "$(get_lastCommitedOpTime ${cluster} ${namespace})" ]]; then
+    if [[ ! -z "$(get_lastCommitedOpTime ${cluster} ${namespace})" \
+          || -n "$(kubectl_bin get deployments -o jsonpath='{.items[?(@.metadata.name == "'"${cluster}-mongos"'")].metadata.name}')" ]]; then
```
Comment (delgod) [inline e2e-tests/rs-shard-migration/run:59]:
```suggestion
    if [[ ! -z "$(get_lastCommitedOpTime ${cluster} ${namespace})" ]] \
          || [[ -n "$(kubectl_bin get deployments -o jsonpath='{.items[?(@.metadata.name == "'"${cluster}-mongos"'")].metadata.name}')" ]]; then
```

**Review by delgod — CHANGES_REQUESTED** [review]

**Review by tplavcic — CHANGES_REQUESTED** [review]
when we move to sharding please check:
- stateful set status for cfg replica
(sorry didn't think about it first time, but seems we don't check it)

after migrating from sharding back to replica please check that:
- `kubectl get svc` doesn't include `some-name-cfg` and `some-name-mongos`
- `kubectl get sts` doesn't include sts for cfg replica

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-497-9d9590aa

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| demand-backup-sharded | passed |
| users | passed |
| upgrade | passed |
| upgrade-sharded | passed |

**Comment by cap1984** [comment]
> when we move to sharding please check:
> 
> * stateful set status for cfg replica
>   (sorry didn't think about it first time, but seems we don't check it)
> 
> after migrating from sharding back to replica please check that:
> 
> * `kubectl get svc` doesn't include `some-name-cfg` and `some-name-mongos`
> * `kubectl get sts` doesn't include sts for cfg replica

done

File: e2e-tests/rs-shard-migration/run (hunk 4 above)
Comment (cap1984) [reply e2e-tests/rs-shard-migration/run]: done

**Review by tplavcic — APPROVED** [review]

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | failed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| demand-backup-sharded | passed |
| users | passed |
| upgrade | passed |
| upgrade-sharded | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| scheduled-backup | failed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| demand-backup-sharded | failed |
| upgrade | passed |
| upgrade-sharded | passed |

**Review by hors — APPROVED** [review]

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-497-d1d6a9e9

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| init-deploy | passed |
| liveness | passed |
| limits | passed |
| smart-update | passed |
| scaling | passed |
| version-service | passed |
| security-context | passed |
| users | passed |
| rs-shard-migration | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| self-healing | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | failed |
| security-context | passed |
| operator-self-healing | passed |
| arbiter | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| demand-backup-sharded | passed |
| version-service | passed |
| upgrade | passed |
| users | passed |
| upgrade-sharded | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-497-ed14fdad

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| limits | passed |
| self-healing | passed |
| scaling | passed |
| operator-self-healing | passed |
| security-context | passed |
| rs-shard-migration | passed |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |

**Pull request merged**